File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
	static int N, M, K;
	static long[] arr, tree;

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		StringBuilder sb = new StringBuilder();

		// 수의 개수 N
		N = Integer.parseInt(st.nextToken());

		// 수의 변경이 일어나는 횟수 M
		M = Integer.parseInt(st.nextToken());

		// 구간 합을 구하는 횟수 K
		K = Integer.parseInt(st.nextToken());

		arr = new long[N + 1];
		tree = new long[N + 1];

		// 값 초기화
		for (int i = 1; i <= N; i++) {
			long input = Long.parseLong(br.readLine());
			update(i, input);
			arr[i] = input;
		}

		for (int i = 0; i < M + K; i++) {
			st = new StringTokenizer(br.readLine());
			int type = Integer.parseInt(st.nextToken());

			int b = Integer.parseInt(st.nextToken());

			if (type == 1) {
				long c = Long.parseLong(st.nextToken());
				update(b, c - arr[b]);
				arr[b] = c;
			} else {
				int c = Integer.parseInt(st.nextToken());
				sb.append(intervalSum(b, c)).append("\n");
			}

		} // for

		System.out.println(sb);
	} // main

	static long prefixSum(int i) {
		long result = 0;
		while (i > 0) {
			result += tree[i];
			i -= (i & -i);
		}
		return result;
	}

	static void update(int i, long val) {
		while (i <= N) {
			tree[i] += val;
			i += (i & -i);
		}
	}

	static long intervalSum(int start, int end) {
		return prefixSum(end) - prefixSum(start - 1);
	}
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.js
// let input = require('fs').readFileSync('/dev/stdin').toString().split('\n');
// 이건 백준에서 제출시 인풋 받기

const init = function (node, start, end) {
    if (start === end) {
        tree[node] = arr[start];
        return;
    }

    let mid = ~~((start + end) / 2);
    init(node * 2, start, mid);
    init(node * 2 + 1, mid + 1, end);
    tree[node] = tree[node * 2] + tree[node * 2 + 1];
}

const update = function (node, start, end, idx, val) {
    if (idx < start || end < idx) {
        return;
    }

    if (start === end) {
        arr[idx] = val;
        tree[node] = val;
        return;
    }

    let mid = ~~((start + end) / 2);
    update(node * 2, start, mid, idx, val);
    update(node * 2 + 1, mid + 1, end, idx, val);
    tree[node] = tree[node * 2] + tree[node * 2 + 1];
}

const query = function (node, start, end, left, right) {
    if (right < start || end < left) {
        return BigInt(0);
    }

    if (left <= start && end <= right) {
        return tree[node];
    }

    let mid = Math.floor(((start + end) / 2));
    let l = query(node * 2, start, mid, left, right);
    let r = query(node * 2 + 1, mid + 1, end, left, right);
    return BigInt(l + r);
}


let input = require('fs').readFileSync('/dev/stdin').toString().split('\n');

let [N, M, K] = input[0].split(' ');
[N, M, K] = [N, M, K].map((element) => {
    return parseInt(element);
})

let arr = [];
for (let i = 0; i < N; ++i) {
    arr.push(BigInt(input[i + 1]));
}

let tree = new Array(4 * N).fill(BigInt(0));

init(1, 0, N - 1);

let ans =[];
for (let i = 1 + N; i < N + M + K + 1; ++i) {
    let [a, b, c] = input[i].split(' ').map((element) => {
        return parseInt(element);
    });

    if (a === 1) {
        c = BigInt(c);
        update(1, 0, N - 1, b - 1, c);
    } else {
        ans.push(query(1, 0, N - 1, b - 1, c - 1).toString(10));
    }
}

console.log(ans.join('\n'));
